src/freertos_drivers/esp_idf/EspIdfWiFi.cxx

```diff
@@ -0,0 +1,1375 @@
+            // There is a very small chance for a race condition whereby another
+            // call "sta_connect()" call has been made with a different password
+            // after this connection has been established. The chances are very
+            // small and the harm is very limited to the next fast connect
+            // attempt failing.
+            std::string ssid((char*)sdata->ssid, sdata->ssid_len);
+            last_sta_update(
+                ssid, staConnectPass_, sdata->authmode, sdata->channel);
+            wlan_connected(true, CONNECT_OK, ssid);
+            break;
+        }
+        case WIFI_EVENT_STA_DISCONNECTED:
+        {
+            // Note: It is possible to get here when we are already
+            //       "disconnected". Therefore, we should not assert on
+            //       connected_ == true. Every failed connection "attempt"
+            //       lands us here.
+            connected_ = false;
+            wifi_event_sta_disconnected_t *evdata =
+                (wifi_event_sta_disconnected_t *)data;
+            LOG(INFO, "wifi: STA disconnected ssid=%s reason %d rssi=%d",
+                evdata->ssid, evdata->reason, evdata->rssi);
+
+            bool try_fast_reconnect = false;
+            // For some reason the IP_EVENT_STA_LOST_IP event does not get
+            // delivered when there is an unexpected WiFi disconnect. I guess
+            // it is implied?
+            if (ipAcquiredSta_)
+            {
+                ipAcquiredSta_ = false; // Disconnect implies we lost IP.
+                try_fast_reconnect = true; // Try a fast reconnect.
+                mdns_disable_sta();
+                ip_acquired(IFACE_STA, false);
+            }
+            if (try_fast_reconnect || fastConnectOnlySta_)
+            {
+                // First reconnect attempt, or forced "fast" connect.
+                esp_wifi_connect();
+                LOG(INFO, "wifi: STA disconnected, fast reconnect attempt...");
+            }
+            else
+            {
+                // Successive reconnect attempts.
+                esp_wifi_scan_start(nullptr, false);
+                LOG(INFO, "wifi: STA disconnected, scanning...");
+            }
+            std::string ssid((char*)evdata->ssid, evdata->ssid_len);
+            wlan_connected(
+                false, connection_result_encode(evdata->reason), ssid);
+            break;
+        }
+        case WIFI_EVENT_SCAN_DONE:
+        {
+            /// @todo The newer IDF versions have a new api that can read
+            ///       AP records one at a time, which is much more memory
+            ///       friendly. Revisit this when updating the IDF version.
+            int idx = -1;
+            uint16_t number;
+            esp_wifi_scan_get_ap_num(&number);
+            LOG(INFO, "wifi: Scan done, number of records: %u.", number);
+            number = std::min(number, static_cast<uint16_t>(10));
```

Comment:
couldwe move this "10" to a static constant?
I remember a situation we encountered in a convention where a hotel had about 20 wifi aps visible and we were unableto connect to LAYOUTWIFI because all we saw in the list was hotel_ap hotel_ap etc.

Reply:
Restricting to 10 maximum is also a potential problem of when it might be applied, from the snippet of context we will display potentially 15 records and the code then will force restrict to 10 maximum. We should defer restriction until after deduplication (hotel case having each floor SSID being identical but different BSSID) and filtering to "interesting" names (LAYOUTWIFI case).